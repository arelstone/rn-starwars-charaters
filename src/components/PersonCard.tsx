import React, { FC } from 'react';
import { StyleSheet, Text, TouchableOpacity, useWindowDimensions, View } from 'react-native';
import { connect } from 'react-redux';
import { RootState } from '../store';
import { Person, Planet } from '../SwApi';
import { NUMBER_OF_COLUMNS } from '../screens/MainScreen';
import colors from '../colors';

interface PersonCardProps {
    charater: Person;
    planet?: Planet;
    onPress: () => void;
}

export const PersonCard: FC<PersonCardProps> = ({ onPress, charater, planet }) => {
    const { width } = useWindowDimensions();
    const dimentions = (width / NUMBER_OF_COLUMNS) - 15;

    return <TouchableOpacity
        onPress={onPress}
        style={[styles.container, { width: dimentions, height: dimentions }]}
    >
        <View style={styles.textContainer}>
            <Text
                style={[styles.text, styles.charaterName]}
                testID="CharaterName"
            >
                {charater.name}
            </Text>
            <Text
                style={styles.text}
                testID="PlanetName"
            >
                {planet?.name || 'N/A'}
            </Text>
        </View>
    </TouchableOpacity>;
};

const mapStateToProps = ({ planets }: RootState, { charater }: PersonCardProps) => ({
    planet: planets.results.find(planet => planet.id === charater.id),
});

export default connect(mapStateToProps)(PersonCard);

const styles = StyleSheet.create({
    container: {
        margin: 5,
        padding: 5,
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: colors.cardBackgroundColor,
    },
    textContainer: {
        alignItems: 'center',
    },
    text: {
        textAlign: 'center',
    },
    charaterName: {
        fontSize: 20,
    },
});
