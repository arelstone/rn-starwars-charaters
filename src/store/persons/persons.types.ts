import { Person } from '../../SwApi';

export const START_FETCHING_PERSONS = 'START_FETCHING_PERSONS';
export const DONE_FETCHING_PERSONS = 'DONE_FETCHING_PERSONS';
export const SET_PERSONS = 'SET_PERSONS';
export const FAILED_FETCHING_PERSONS = 'FAILED_FETCHING_PERSONS';

export type SetPersonsParams = {
    count: number;
    nextPage: number;
    results: Person[];
}

export type FetchPersonsAction = {
    type: PersonActionTypes;
}

type PersonActionTypes = typeof START_FETCHING_PERSONS
    | typeof DONE_FETCHING_PERSONS
    | typeof FAILED_FETCHING_PERSONS;

export type SetPersonsAction = {
    type: typeof SET_PERSONS;
    payload: SetPersonsParams;
}
