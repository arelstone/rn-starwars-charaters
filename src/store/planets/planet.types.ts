import { Planet } from '../../SwApi';

export const START_FETCHING_PLANETS = 'START_FETCHING_PLANETS';
export const DONE_FETCHING_PLANETS = 'DONE_FETCHING_PLANETS';
export const SET_PLANETS = 'SET_PLANETS';
export const FAILED_FETCHING_PLANETS = 'FAILED_FETCHING_PLANETS';

export type SetPlanetsParams = {
    count: number;
    nextPage: number;
    results: Planet[];
}

type PlanetActionTypes = typeof SET_PLANETS
    | typeof DONE_FETCHING_PLANETS
    | typeof START_FETCHING_PLANETS
    | typeof FAILED_FETCHING_PLANETS;

export type FetchPlanetsAction = {
    type: PlanetActionTypes;
}
