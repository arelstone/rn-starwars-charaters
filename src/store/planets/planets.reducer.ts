import { Planet } from '../../SwApi';
import { START_FETCHING_PLANETS, DONE_FETCHING_PLANETS, SET_PLANETS, FAILED_FETCHING_PLANETS } from './planet.types';
import { StoreResource } from '../root.reducer';


export const INITIAL_STATE: StoreResource<Planet[]> = {
    count: 0,
    loading: false,
    nextPage: undefined,
    results: [],
    error: {
        hasError: false,
        message: '',
    },
};

export default (state = INITIAL_STATE, { type, payload }: any): StoreResource<Planet[]> => {
    switch (type) {
        case START_FETCHING_PLANETS:
            return {
                ...state,
                results: [],
                loading: true,
                error: {
                    hasError: false,
                    message: '',
                },
            };
        case DONE_FETCHING_PLANETS:
            return {
                ...state,
                loading: false,
            };
        case FAILED_FETCHING_PLANETS:
            return {
                ...state,
                loading: false,
                error: {
                    hasError: true,
                    message: payload,
                },
            };
        case SET_PLANETS: {
            return {
                ...state,
                ...payload,
                results: [
                    ...state.results,
                    ...payload.results,
                ],
            };
        }
        default:
            return state;
    }
};

