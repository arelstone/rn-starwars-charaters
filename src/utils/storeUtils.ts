export const getNumber = (string: string) => Number(string.replace(/\D/g, ''));
