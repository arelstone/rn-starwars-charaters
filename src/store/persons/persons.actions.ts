import { Person, ResourceType } from '../../SwApi';
import { getNumber, nextPage } from '../../utils/storeUtils';
import { FAILED_FETCHING_PERSONS, START_FETCHING_PERSONS, SET_PERSONS, DONE_FETCHING_PERSONS } from './persons.types';
import { baseUrl } from '../../config';

export const fetchPersons = (page = 1, uri: ResourceType = ResourceType.People) => async (dispatch: any) => {
    if (page === 1) {
        dispatch({ type: START_FETCHING_PERSONS });
    }

    const response = await fetch(`${baseUrl()}/${uri}/?page=${page}`)
        .then(r => r.json())
        .catch(error => dispatch({
            type: FAILED_FETCHING_PERSONS,
            payload: error.message,
        }));

    const next = nextPage(response);

    dispatch({
        type: SET_PERSONS,
        payload: {
            nextPage: next,
            count: response.count,
            results: response.results.map((person: Person) => ({
                ...person,
                id: getNumber(person.url),
                homeworld_id: getNumber(person.homeworld),
            }))
                .sort((a: Person, b: Person) => a.name.toLowerCase().localeCompare(b.name.toLowerCase())),
        },
    });

    if (next) {
        return dispatch(fetchPersons(next));
    }
    console.log('###HERE', next);


    return dispatch({ type: DONE_FETCHING_PERSONS });
};
