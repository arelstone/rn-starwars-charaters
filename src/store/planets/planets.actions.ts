import { Dispatch } from 'react';
import { GetState } from '..';
import { Planet } from '../../SwApi';
import { getNumber } from '../../utils/storeUtils';
import { SetPlanetsParams, FetchPlanetsAction } from './planet.types';

export const fetchPlanets = (page = 1) => async (dispatch: Dispatch<FetchPlanetsAction>, getState: GetState) => {
    if (page === 1) {
        dispatch({ type: 'START_FETCHING_PLANETS' });
    }
    const response = await fetch(`https://swapi.dev/api/planets/?page=${page}`)
        .then(r => r.json())
        .catch(error => console.log('###error', error));

    const nextPage = response.next && getNumber(response.next);

    const results = response.results.map((result: Planet) => ({
        ...result,
        id: getNumber(result.url),
    }));

    dispatch(setPlanets({
        nextPage: response.next && getNumber(response.next),
        results: [...results, ...getState().planets.results],
        count: response.count,
    }));

    // if (nextPage) {
    //     return dispatch(fetchPlanets(nextPage));
    // }

    return dispatch({ type: 'DONE_FETCHING_PLANETS' });
};

const setPlanets = (payload: SetPlanetsParams) => dispatch => {
    return dispatch({
        type: 'SET_PLANETS',
        payload,
    });
};
