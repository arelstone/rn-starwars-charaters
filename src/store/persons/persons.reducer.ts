import { Person } from '../../SwApi';
import { StoreResource } from '../root.reducer';
import { DONE_FETCHING_PERSONS, FAILED_FETCHING_PERSONS, SET_PERSONS, START_FETCHING_PERSONS } from './persons.types';

export const INITIAL_STATE: StoreResource<Person[]> = {
    count: 0,
    loading: false,
    error: {
        hasError: false,
        message: '',
    },
    nextPage: undefined,
    results: [],
};

export default (state = INITIAL_STATE, { type, payload }: any): StoreResource<Person[]> => {
    switch (type) {
        case START_FETCHING_PERSONS:
            return {
                ...state,
                results: [],
                loading: true,
                error: {
                    hasError: false,
                    message: '',
                },
            };
        case DONE_FETCHING_PERSONS:
            return {
                ...state,
                loading: false,
            };
        case FAILED_FETCHING_PERSONS:
            return {
                ...state,
                loading: false,
                error: {
                    hasError: true,
                    message: payload,
                },
            };
        case SET_PERSONS: {
            return {
                ...state,
                ...payload,
                results: [
                    ...state.results,
                    ...payload.results,
                ],
            };
        }
        default:
            return state;
    }
};

