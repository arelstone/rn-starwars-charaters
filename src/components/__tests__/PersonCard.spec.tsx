import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { PersonCard } from '../PersonCard';
import { personsMock } from '../../../__mocks__/persons.mock';
import { planetsMock } from '../../../__mocks__/planets.mock';
import { findElementByTestId } from '../../utils/testUtils';

let wrapper: ShallowWrapper;
const mockOnPress = jest.fn();

describe('<PersonCard />', () => {
    beforeAll(() => {
        wrapper = shallow(<PersonCard
            charater={personsMock[0]}
            onPress={mockOnPress}
            planet={planetsMock[0]}
        />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render the CharaterName', () => {
        expect(findElementByTestId('Text', 'CharaterName', wrapper).prop('children')).toEqual(personsMock[0].name);
    });

    it('should render the PlanetName', () => {
        expect(findElementByTestId('Text', 'PlanetName', wrapper).prop('children')).toEqual(planetsMock[0].name);
    });

    it('should invoke onPress', () => {
        // eslint-disable-next-line
        // @ts-expect-error
        wrapper.prop('onPress')();

        expect(mockOnPress).toHaveBeenCalled();
    });
});
