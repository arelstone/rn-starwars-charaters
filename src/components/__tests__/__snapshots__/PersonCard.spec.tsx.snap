// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PersonCard /> should match snapshot 1`] = `
<ForwardRef
  onPress={[MockFunction]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0.2)",
        "justifyContent": "center",
        "margin": 5,
        "padding": 5,
      },
      Object {
        "height": 235,
        "width": 235,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "fontSize": 20,
          },
        ]
      }
      testID="CharaterName"
    >
      Luke Skywalker
    </Text>
    <Text
      style={
        Object {
          "textAlign": "center",
        }
      }
      testID="PlanetName"
    >
      Tatooine
    </Text>
  </View>
</ForwardRef>
`;
